### PR TITLE
[docs] Update `checkpointing.rst` and `callbacks.rst` for Stateful support

### LINKED ISSUE #11429: [RFC] Standardize all stateful components on `state_dict`/`load_state_dict`
## Proposed refactor
Standardize all stateful components on `state_dict`/`load_state_dict`.

### Background

PyTorch convention uses `state_dict`/`load_state_dict` for gathering and loading object state. In lightning, some components follow this convention, while other components do not (see Appedix: current state gathering section)

### Motivation
 
Each component should contribute saving/loading their own state with the same APIs.

1. independently contributing/loading one’s own local component state (aligning with PyTorch primitives: `state_dict`/`load_state_dict`)
2. operating and depending on global component state (Lightning `CheckpointHooks`: `on_save/load_checkpoint`)

**This issue is focused on aligning all components on 1**. Following this convention will allow consistency across Lightning components and consistency with PyTorch conventions. Any stateful component can simply implement their own `state_dict`/`load_state_dict` methods to contribute their own state.
For now **2** is only adjusted as needed (Callbacks).

### Pitch

Lightning already has this Stateful Protocol in `auto_restart.py`. We can move this `_SupportsStateDict` out to the more central `core/hooks.py` file:

https://github.com/PyTorchLightning/pytorch-lightning/blob/34c62da37dc1ed9a1de7e023c690c7528ee56c60/pytorch_lightning/utilities/auto_restart.py#L638-L646

```py
@runtime_checkable
class Stateful(Protocol):
    """This class is used to detect if an object is stateful using `isinstance(obj, Stateful)`."""

    def state_dict(self) -> Dict[str, Any]:
        ...

    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
        ...
```
### Additional context

Part of #7740

### Appendix: current state gathering:
Specifically, current components contribute state in the following different ways:
1. Some components contribute saving/loading their own state with only `on_save/load_checkpoint` [DataModule, PrecisionPlugin, Callbacks]
a. [DataModule, PrecisionPlugin] use `on_save/load_checkpoint` from `CheckpointHooks` https://github.com/PyTorchLightning/pytorch-lightning/blob/948cfd24de4f64a2980395581f15544e5e37eab0/pytorch_lightning/core/hooks.py#L765-L807
b. [Callbacks] contribute saving/loading their own state with different `on_save/load_checkpoint` hook signatures and functionality https://github.com/PyTorchLightning/pytorch-lightning/blob/948cfd24de4f64a2980395581f15544e5e37eab0/pytorch_lightning/callbacks/base.py#L293-L323
c. though [Loops] falls under 3. below, noting here that Loops also have their own `on_save/load_checkpoint` methods with different signatures. https://github.com/PyTorchLightning/pytorch-lightning/blob/948cfd24de4f64a2980395581f15544e5e37eab0/pytorch_lightning/loops/base.py#L253-L262
2. Some components contribute saving/loading their own state with only `state_dict`/`load_state_dict` calls [Optimizers, LR schedulers]
3. Some components have both [LightningModule, Loops]
#### Save aligning on state_dict
`dump_checkpoint`
https://github.com/PyTorchLightning/pytorch-lightning/blob/06b8f82b8a97e7e5653486731751521478f58cce/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L310-L393
becomes:
```py
def dump_checkpoint(self, weights_only: bool = False) -> dict:
   ...
   # dump callbacks
   # checkpoint["callbacks"] = self.trainer._call_callbacks_on_save_checkpoint(checkpoint)
   # becomes
   checkpoint["callbacks"] = self.trainer._call_callbacks_state_dict()
   
   ...
   # precision plugin
   # self.trainer.precision_plugin.on_save_checkpoint(checkpoint)
   # becomes
   prec_plugin = self.trainer.precision_plugin
   checkpoint[prec_plugin.__class__.__name__] = self.trainer.precision_plugin.state_dict()

   ...
   # give the model a chance to dump a few things
   # model.on_save_checkpoint(checkpoint)
   # if self.trainer.datamodule is not None:
   #    self.trainer.datamodule.on_save_checkpoint(checkpoint)
   # becomes
   
   # datamodule state
   dm = self.trainer.datamodule
   if dm is not None:
       checkpoint[dm.__class__.__name__] = dm.state_dict()
       
   # on_save_checkpoint calls
   model.on_save_checkpoint(checkpoint)
   dm.on_save_checkpoint(checkpoint)
   prec_plugin.on_save_checkpoint(checkpoint)
   for callback in self.trainer.callbacks:
       callback.on_save_checkpoint(self.trainer, model, checkpoint)
```
#### Load aligning on load_state_dict
see component Load sections below

#### Base Class

https://github.com/PyTorchLightning/pytorch-lightning/blob/59a7ba760548baadf6dbb30864b54cb01c7225a3/pytorch_lightning/callbacks/base.py#L293-L323

becomes
BC: `on_save_checkpoint` returns `None` instead of `dict`
BC: `on_load_checkpoint` arg takes entire `checkpoint` dict instead of `callback_state`
```py
def state_dict(self) -> Dict[str, Any]:
    return {}

def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
    pass

def on_save_checkpoint(
        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", checkpoint: Dict[str, Any]
    ) -> None:
    """Called by Lightning when saving a checkpoint to give you a chance to store or customize anything
    else you might want to save.
    Args:
        trainer: the current :class:`~pytorch_lightning.trainer.Trainer` instance.
        pl_module: the current :class:`~pytorch_lightning.core.lightning.LightningModule` instance.
        checkpoint: the checkpoint dictionary that will be saved.
    """
    pass

def on_load_checkpoint(
    self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", checkpoint: Dict[str, Any]
) -> None:
    """Called by Lightning when loading a checkpoint to give you a chance to reload or customize anything
    else you may have saved in on_save_checkpoint.
    Args:
        trainer: the current :class:`~pytorch_lightning.trainer.Trainer` instance.
        pl_module: the current :class:`~pytorch_lightning.core.lightning.LightningModule` instance.
        checkpoint: entire loaded checkpoint dictionary
    """
    pass
```
#### Save
https://github.com/PyTorchLightning/pytorch-lightning/blob/85304d4672a9ed24a16f7f5b2abaa34148ab86f4/pytorch_lightning/trainer/trainer.py#L1603
```py
def _call_callbacks_on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[str, dict]:
    ...
    state = callback.on_save_checkpoint(self, self.lightning_module, checkpoint)
```
becomes
```py
def _call_callbacks_state_dict(self) -> Dict[str, dict]:
   ...
   state = callback.state_dict()
```
#### Load
https://github.com/PyTorchLightning/pytorch-lightning/blob/06b8f82b8a97e7e5653486731751521478f58cce/pytorch_lightning/trainer/trainer.py#L1652
```py
def _call_callbacks_on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
    ...
    callback.on_load_checkpoint(self, self.lightning_module, state)
```
becomes
```py
def _call_callbacks_load_state_dict(self, checkpoint: Dict[str, Any]) -> None:
    ...
    callback.load_state_dict(state)
```
`restore_callbacks`
https://github.com/PyTorchLightning/pytorch-lightning/blob/06b8f82b8a97e7e5653486731751521478f58cce/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L205
becomes
```py
for callback in self.trainer.callbacks:
    callback.on_load_checkpoint(self.trainer, self.trainer.lightning_module, self._loaded_checkpoint)
self.trainer._call_callbacks_load_state_dict(self._loaded_checkpoint)
```
#### Callback classes
update `timer`, `pruning`, `model_checkpoint`, `finetuning`, `early_stopping` to use `state_dict`/`load_state_dict` instead of `on_save/load_checkpoint`.

#### Base Class
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/precision/precision_plugin.py
add dummy `state_dict/load_state_dict`:
```py
def state_dict(self) -> Dict[str, Any]:
    return {}
def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
    pass
```
#### Load
`restore_training_state`
https://github.com/PyTorchLightning/pytorch-lightning/blob/06b8f82b8a97e7e5653486731751521478f58cce/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L189-L190
becomes
```py
# restore precision plugin (scaler etc.)
prec_plugin = self.trainer.precision_plugin
prec_plugin.on_load_checkpoint(self._loaded_checkpoint)
if prec_plugin.__class__.__name__ in self._loaded_checkpoint:
    prec_plugin.load_state_dict(self._loaded_checkpoint[prec_plugin.__class__.__name__])
```
#### Precision Plugin classes

Update `apex_amp`, `native_amp` to use `state_dict/load_state_dict` instead of `on_save/load_checkpoint`

#### Base Class
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/datamodule.py
add dummy `state_dict/load_state_dict`:
```py
def state_dict(self) -> Dict[str, Any]:
    return {}
def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
    pass
```
#### Load

`restore_datamodule`
https://github.com/PyTorchLightning/pytorch-lightning/blob/06b8f82b8a97e7e5653486731751521478f58cce/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L151
becomes
```py
datamodule.on_load_checkpoint(self._loaded_checkpoint)
if datamodule.__class__.__name__ in self._loaded_checkpoint:
    datamodule.load_state_dict(self._loaded_checkpoint[datamodule.__class__.__name__])
```

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @rohitgr7 @ananthsub @ninginthecloud

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

docs for : Fixes #11429

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/extensions/callbacks.rst [hunk 1]
```diff
@@ -422,12 +422,24 @@ on_exception
 .. automethod:: pytorch_lightning.callbacks.Callback.on_exception
     :noindex:
 
+state_dict
+~~~~~~~~~~~~~~~~~~
+
+.. automethod:: pytorch_lightning.callbacks.Callback.state_dict
+    :noindex:
+
 on_save_checkpoint
```
Comment (rohitgr7) [inline docs/source/extensions/callbacks.rst:431]: same here. we should remove the explicit mention to reduce the visibility. From an API point of view, they are already present in the API references.

File: docs/source/extensions/callbacks.rst [hunk 2]
```diff
@@ -422,12 +422,24 @@ on_exception
...
 .. automethod:: pytorch_lightning.callbacks.Callback.on_save_checkpoint
     :noindex:
 
+load_state_dict
+~~~~~~~~~~~~~~~~~~
+
+.. automethod:: pytorch_lightning.callbacks.Callback.load_state_dict
+    :noindex:
+
 on_load_checkpoint
```
Comment (rohitgr7) [inline docs/source/extensions/callbacks.rst:443]: same here.

File: docs/source/common/checkpointing.rst [hunk 3]
```diff
@@ -28,11 +28,27 @@ A Lightning checkpoint has everything needed to restore a training session inclu
...
+===========================
+
+Each component can save and load their state by implementing PyTorch ``state_dict``, ``load_state_dict`` stateful protocol.
+For details on implementing your own stateful callbacks and datamodules, reference their docs pages at :doc:`callbacks <../extensions/callbacks>` and :doc:`datamodules <../extensions/datamodules>`.
+
+
+Operating on Global Checkpoint Component States
+===============================================
+
+If you need to operate on the global component state (i.e. the entire checkpoint dictionary), you can add/delete/modify custom states in your checkpoints before they are being saved or loaded.
```
Comment (rohitgr7) [inline docs/source/common/checkpointing.rst:47]:
```suggestion
If you need to operate on the global component state (i.e. the entire checkpoint dictionary), you can read/add/delete/modify custom states in your checkpoints before they are being saved or loaded.
```

File: docs/source/common/checkpointing.rst [hunk 4]
```diff
@@ -28,11 +28,27 @@ A Lightning checkpoint has everything needed to restore a training session inclu
...
+- State of all callbacks (for stateful callbacks)
+- State of datamodule (for stateful datamodules)
 - The hyperparameters used for that model if passed in as hparams (Argparse.Namespace)
 - State of Loops (if using Fault-Tolerant training)
 
 
+Individual Component States
+===========================
+
+Each component can save and load their state by implementing PyTorch ``state_dict``, ``load_state_dict`` stateful protocol.
```
Comment (rohitgr7) [inline docs/source/common/checkpointing.rst:40]:
```suggestion
Each component can save and load its state by implementing PyTorch ``state_dict``, ``load_state_dict`` stateful protocol.
```

**Review by rohitgr7 — APPROVED** [review]

File: docs/source/extensions/callbacks.rst (hunk 1 above)
Comment (krshrimali) [reply docs/source/extensions/callbacks.rst:431]: Thinking out loud: Should we still mention but also say that this is deprecated?

Comment (rohitgr7) [reply docs/source/extensions/callbacks.rst:431]: anyway is better. I usually remove it since if it's a new user, they better not know about this and if it's an old user, probably they will search for it and API references will pop up for them.

File: docs/source/common/checkpointing.rst [hunk 5]
```diff
@@ -28,11 +28,27 @@ A Lightning checkpoint has everything needed to restore a training session inclu
...
+For details on implementing your own stateful callbacks and datamodules, reference their docs pages at :doc:`callbacks <../extensions/callbacks>` and :doc:`datamodules <../extensions/datamodules>`.
+
+
+Operating on Global Checkpoint Component States
+===============================================
+
+If you need to operate on the global component state (i.e. the entire checkpoint dictionary), you can add/delete/modify custom states in your checkpoints before they are being saved or loaded.
+For this you can override :meth:`~pytorch_lightning.core.hooks.CheckpointHooks.on_save_checkpoint` and :meth:`~pytorch_lightning.core.hooks.CheckpointHooks.on_load_checkpoint` in your ``LightningModule``
+or :meth:`~pytorch_lightning.callbacks.base.Callback.on_save_checkpoint` and :meth:`~pytorch_lightning.callbacks.base.Callback.on_load_checkpoint` methods in your ``Callback``.
```
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply docs/source/common/checkpointing.rst:49]: they're not deprecated, but rather the signature will change in 1.8 for callbacks (`on_save_checkpoint -> dict` becomes `on_save_checkpoint -> None`, `on_load_checkpoint(callback_state)` becomes `on_load_checkpoint(checkpoint_dict)`). I think it's ok to leave here bc the docstring will indicate the changes (see #11887)

File: docs/source/extensions/callbacks.rst (hunk 1 above)
Comment (jjenniferdai) [reply docs/source/extensions/callbacks.rst:431]: (see above [comment](https://github.com/PyTorchLightning/pytorch-lightning/pull/12351#discussion_r829276479))

File: docs/source/extensions/callbacks.rst (hunk 2 above)
Comment (jjenniferdai) [reply docs/source/extensions/callbacks.rst:443]: (same, see above [comment](https://github.com/PyTorchLightning/pytorch-lightning/pull/12351#discussion_r829276479))

**Review by ananthsub — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
great

File: docs/source/extensions/callbacks.rst [hunk 6]
```diff
@@ -422,12 +422,24 @@ on_exception
 .. automethod:: pytorch_lightning.callbacks.Callback.on_exception
     :noindex:
 
+state_dict
+~~~~~~~~~~~~~~~~~~
```
Comment (Borda) [inline docs/source/extensions/callbacks.rst:426]:
```suggestion
~~~~~~~~~~
```

File: docs/source/extensions/callbacks.rst [hunk 7]
```diff
@@ -422,12 +422,24 @@ on_exception
...
+    :noindex:
+
 on_save_checkpoint
 ~~~~~~~~~~~~~~~~~~
 
 .. automethod:: pytorch_lightning.callbacks.Callback.on_save_checkpoint
     :noindex:
 
+load_state_dict
+~~~~~~~~~~~~~~~~~~
```
Comment (Borda) [inline docs/source/extensions/callbacks.rst:438]:
```suggestion
~~~~~~~~~~~~~~~
```

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
